src/project.coffee

```diff
@@ -182,22 +190,22 @@ class Project extends Model
     Grim.deprecate("Use ::setPaths instead")
     @setPaths([path])
 
-  # Public: Add a path the project's list of root paths
+  # Public: Add a path to the project's list of root paths
   #
   # * `projectPath` {String} The path to the directory to add.
   addPath: (projectPath, options) ->
-    projectPath = path.normalize(projectPath)
-
-    directoryPath = if fs.isDirectorySync(projectPath)
-      projectPath
-    else
-      path.dirname(projectPath)
-
-    return if @getPaths().some (existingPath) ->
-      (directoryPath is existingPath) or
-      (directoryPath.indexOf(path.join(existingPath, path.sep)) is 0)
-
-    directory = new Directory(directoryPath)
+    for directory in @getDirectories()
+      # Apparently a Directory does not believe it can contain itself, so we
+      # must also check whether the paths match.
+      return if directory.contains(projectPath) or directory.getPath() is projectPath
+
+    directory = null
+    for provider in @directoryProviders
+      break if directory = provider.directoryForURISync?(projectPath)
+    if directory is null
+      # This should never happen because DefaultDirectoryProvider should always
+      # return a Directory.
```

Comment:
I still think that if an unrecognized URI is given, this *should* happen. The `DefaultDirectoryProvider` should return `null` unless the URI is for a local file. That way, this would actually be a useful error.

I also still think the `DefaultDirectoryProvider` should be tried first, not last, and that it should simply be called `LocalDirectoryProvider`.